x/cash/msg_test.go

```diff
@@ -1,138 +1,149 @@
+				Ref:    []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaB"),
+			},
+			wantErr: errors.ErrInvalidState,
+		},
 	}
-	err = minus.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidAmount.Is(err))
 
-	bad := coin.NewCoin(3, 4, "fab9")
-	ugly := &SendMsg{
-		Amount: &bad,
-		Dest:   addr2,
-		Src:    addr3,
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			if err := tc.msg.Validate(); !tc.wantErr.Is(err) {
+				t.Fatalf("unexpected error: %+v", err)
+			}
+		})
 	}
-	err = ugly.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrCurrency.Is(err))
-
 }
 
 func TestValidateFeeTx(t *testing.T) {
-	var empty *FeeInfo
-	err := empty.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidInput.Is(err))
-
-	addr := weave.NewAddress([]byte{8, 8})
-	addr2 := weave.NewAddress([]byte{7, 7})
-
-	nofee := &FeeInfo{Payer: addr}
-	err = nofee.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidAmount.Is(err))
-
-	pos := coin.NewCoin(10, 0, "FOO")
-	plus := &FeeInfo{Fees: &pos}
-	err = plus.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidInput.Is(err))
-
-	full := plus.DefaultPayer(addr)
-	assert.NoError(t, full.Validate())
-	assert.EqualValues(t, addr, full.GetPayer())
-
-	full2 := full.DefaultPayer(addr2)
-	assert.NoError(t, full2.Validate())
-	assert.EqualValues(t, addr, full2.GetPayer())
-
-	zero := &FeeInfo{
-		Payer: addr2,
-		Fees:  &coin.Coin{Ticker: "BAR"},
+	addr1 := randomAddr(t)
+
+	cases := map[string]struct {
+		info    *FeeInfo
+		wantErr *errors.Error
+	}{
+		"success": {
+			info: &FeeInfo{
+				Fees:  coin.NewCoinp(1, 0, "IOV"),
+				Payer: addr1,
+			},
+			wantErr: nil,
+		},
+		"empty": {
+			info:    &FeeInfo{},
+			wantErr: errors.ErrInvalidAmount,
+		},
+		"no fee": {
+			info: &FeeInfo{
+				Payer: addr1,
+			},
+			wantErr: errors.ErrInvalidAmount,
+		},
+		"no payer": {
+			info: &FeeInfo{
+				Fees: coin.NewCoinp(10, 0, "IOV"),
+			},
+			wantErr: errors.ErrEmpty,
+		},
+		"negative fee": {
+			info: &FeeInfo{
+				Fees:  coin.NewCoinp(-10, 0, "IOV"),
+				Payer: addr1,
+			},
+			wantErr: errors.ErrInvalidAmount,
+		},
+		"invalid fee ticker": {
+			info: &FeeInfo{
+				Fees:  coin.NewCoinp(10, 0, "foobar"),
+				Payer: addr1,
+			},
+			wantErr: errors.ErrCurrency,
+		},
 	}
-	assert.NoError(t, zero.Validate())
 
-	neg := coin.NewCoin(-3, 0, "FOO")
-	minus := &FeeInfo{
-		Payer: addr,
-		Fees:  &neg,
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			if err := tc.info.Validate(); !tc.wantErr.Is(err) {
+				t.Fatalf("unexpected error: %+v", err)
+			}
+		})
 	}
-	err = minus.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidAmount.Is(err))
+}
 
-	bad := coin.NewCoin(3, 0, "fab9")
-	ugly := &FeeInfo{
-		Payer: addr,
-		Fees:  &bad,
+func randomAddr(t testing.TB) weave.Address {
```

Comment:
pp: you can replace this by `weavetest.NewCondition().Address()` or `weavetest.NewKey().PublicKey().Address()` or just move this into weavetest package

Reply:
https://github.com/iov-one/weave/pull/526/commits/4586c56268bbbb8130b2767b0f1796214ff038ca